internal/executor/rpc/rpc.go

```diff
@@ -0,0 +1,273 @@
+
+	task, err := r.build.GetTaskFromIdentification(req.TaskIdentification, r.clientSecret)
+	if err != nil {
+		return nil, err
+	}
+
+	return &api.CommandsResponse{
+		Environment:      r.build.Environment,
+		Commands:         task.ProtoTask.Commands,
+		ServerToken:      r.serverSecret,
+		TimeoutInSeconds: int64(task.Timeout.Seconds()),
+	}, nil
+}
+
+func getCommandAfter(commands []*api.Command, currentCommand string) string {
+	var doReturn bool
+
+	for _, command := range commands {
+		if doReturn {
+			return command.Name
+		}
+
+		if command.Name == currentCommand {
+			doReturn = true
+		}
+	}
+
+	return ""
+}
+
+func (r *RPC) ReportSingleCommand(
+	ctx context.Context,
+	req *api.ReportSingleCommandRequest,
+) (*api.ReportSingleCommandResponse, error) {
+	r.build.Mutex.Lock()
+	defer r.build.Mutex.Unlock()
+
+	task, err := r.build.GetTaskFromIdentification(req.TaskIdentification, r.clientSecret)
+	if err != nil {
+		return nil, err
+	}
+
+	var nextCommand string
+
+	// Register whether the current command succeeded or failed
+	// so that the main loop can make the decision whether
+	// to proceed with the execution or not.
+	if req.Succeded {
+		nextCommand = getCommandAfter(task.ProtoTask.Commands, req.CommandName)
+		if nextCommand == "" {
+			task.Status = build.StatusSucceeded
+		}
+	} else {
+		// An empty string instructs the agent to do nothing and terminate
+		nextCommand = ""
+		task.Status = build.StatusFailed
+	}
+
+	return &api.ReportSingleCommandResponse{
+		NextCommandName: nextCommand,
+	}, nil
+}
+
+func (r *RPC) StreamLogs(stream api.CirrusCIService_StreamLogsServer) error {
+	var currentTaskName string
+	var currentCommand string
+
+	for {
+		logEntry, err := stream.Recv()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			r.logger.WithContext(stream.Context()).Warn(err)
+			return err
+		}
+
+		switch x := logEntry.Value.(type) {
+		case *api.LogEntry_Key:
+			r.build.Mutex.Lock()
+
+			task, err := r.build.GetTaskFromIdentification(x.Key.TaskIdentification, r.clientSecret)
+			if err != nil {
+				return err
+			}
+			currentTaskName = task.ProtoTask.Name
+			currentCommand = x.Key.CommandName
+
+			r.build.Mutex.Unlock()
+
+			r.logger.WithFields(map[string]interface{}{
+				"task":    currentTaskName,
+				"command": currentCommand,
+			}).Debug("begin streaming logs")
+		case *api.LogEntry_Chunk:
+			if currentTaskName == "" {
+				return status.Error(codes.PermissionDenied, "not authenticated")
+			}
+
+			r.logger.WithFields(map[string]interface{}{
+				"task":    currentTaskName,
+				"command": currentCommand,
+			}).Debugf("received chunk: %s", string(x.Chunk.Data))
+		}
+	}
+
+	if err := stream.SendAndClose(&api.UploadLogsResponse{}); err != nil {
+		r.logger.WithContext(stream.Context()).WithError(err).Warn("while closing log stream")
+		return err
+	}
+
+	return nil
+}
+
+func (r *RPC) Heartbeat(ctx context.Context, req *api.HeartbeatRequest) (*api.HeartbeatResponse, error) {
```

Comment:
Let's log a heartbeat here. Can be useful.

Reply:
Added in https://github.com/cirruslabs/cirrus-cli/pull/6/commits/7fafcd8aaf9c132b30316712571baabc6ee9edc6.